examples/field_examples/viking_graben_line12/fwi/fwi.jl

```diff
@@ -0,0 +1,278 @@
+    @info "save_results: $h5_file"
+    n = (length(x),length(z))
+    o = (x[1],z[1])
+    d = (x[2]-x[1],z[2]-z[1])
+    fwi_plt = Plots.heatmap(x, z, sqrt.(1f0 ./ adjoint(reshape(m,n))), c=:rainbow, 
+        xlims=(x[1],x[end]), 
+        ylims=(z[1],z[end]), yflip=true,
+        title=v_plot_title,
+        xlabel="Lateral position [km]",
+        ylabel="Depth [km]",
+        dpi=600)
+    Plots.savefig(fwi_plt, v_file)
+
+    grad_plt = Plots.heatmap(x, z, adjoint(reshape(grad,n)), c=:bluesreds, 
+        xlims=(x[1],x[end]), 
+        ylims=(z[1],z[end]), yflip=true,
+        clim=(-maximum(grad)/5f0, maximum(grad)/5f0),
+        title=grad_plot_title,
+        xlabel="Lateral position [km]",
+        ylabel="Depth [km]",
+        dpi=600)
+    Plots.savefig(grad_plt,grad_file)
+
+    h5open(h5_file, "w") do file
+        write(file, 
+            "v", sqrt.(1f0 ./ reshape(m, n)), 
+            "m", reshape(m, n), 
+            "o", collect(o.*1000f0), 
+            "n", collect(n), 
+            "d", collect(d.*1000f0),
+            "fhistory", fhistory,
+            "grad", reshape(grad, n))
+    end
+end
+
+# NLopt objective function
+function nlopt_obj_fun!(m_update, grad)
+
+    global x, z, count, jopt, gradientCell, seabed_ind;
+    count += 1
+
+    # Update model
+    model0.m .= Float32.(reshape(m_update, model0.n))
+    if modeling_type == "acoustic"
+        model0.b .= Float32.(reshape(1f0./rho_from_slowness(model0.m), model0.n))
+        model0.b[:,1:seabed_ind] .= 1f0/rhowater
+    end
+
+    # Select batch and calculate gradient
+    # Subsampling the number of sources should in practice never be used for second order methods such as L-BFGS.
+    indsrc = shot_from:shot_step:shot_to
+    if mute_reflections
+        fval, gradient = fwi_objective(model0, q[indsrc], Ml_tur[indsrc]*d_obs[indsrc], options=jopt)
+    elseif mute_turning
+        fval, gradient = fwi_objective(model0, q[indsrc], Ml_ref[indsrc]*d_obs[indsrc], options=jopt)
+    else
+        fval, gradient = fwi_objective(model0, q[indsrc], d_obs[indsrc], options=jopt)
+    end
+    gradient = reshape(gradient, model0.n)
+    gradient[:, 1:seabed_ind] .= 0f0
+
+    grad[1:end] .= gradient[1:end]
+
+    push!(modelCell, copy(m_update))
+    push!(gradientCell, copy(gradient))
+    push!(fhistory, fval)
+
+    println("iteration: ", count, "\tfval: ", fval, "\tnorm: ", norm(gradient))
+    save_results(x,z,model0.m.data,gradient.data,fhistory,
+                "FWI with L-BFGS $modeling_type: $(frq*1000)Hz, iter $count",
+                "Gradient with L-BFGS $modeling_type: $(frq*1000)Hz, iter $count",
+                dir_out * "FWI $count.png",
+                dir_out * "Gradient $count.png",
+                dir_out * model_file_out * " " * string(count) * ".h5")
+
+    return convert(Float64, fval)
+end
+
+println("No.  ", "fval         ", "norm(gradient)")
+
+opt = Opt(:LD_LBFGS, prod(model0.n))
```

Comment:
Any reason you chose this one over something like `pqn` that (in theory) handle constraints like bounds a bit better.

Just a question no need to change it

Reply:
Please share a link to `pqn`. Don't know about that.